### ISSUE TITLE
[path_provider] Add support for web

### ISSUE THREAD

**Issue by hterkelsen** [flutter/flutter#45296]
We should add web support for `path_provider`

**Comment by MichelFR** [comment]
Since its not possible to access the filessystem on the web...  how should this be done?

**Comment by jld3103** [comment]
I'm experiencing a performance issue with the same error as https://github.com/mchome/flutter_advanced_networkimage/issues/144. I'll try to create a stub implementation for it (and also migrate it to the platform interface).

**Comment by TimWhiting** [comment]
I can imagine that this might be useful in some contexts on the web. Like which folder to open a file-download picker to. (For user content). For cached content, I'm not sure it makes sense on the web. But it is pretty annoying to get a flood of errors in the console on web when trying to use a package like google_fonts which uses this plugin.

**Comment by jld3103** [comment]
I actually fixed it for the google_fonts package some days ago. Just update. But still a lot of packages use it like on mobile.

**Comment by ditman** [comment]
Some work related to federation is happening here:

https://github.com/flutter/flutter/issues/46307#issuecomment-589301237

**Comment by simmi9** [comment]
I was trying to download a file on my Windows OS using path_provider. getDownloadsDirectory(), and faced issues related to platform, then found this existing issue link. Looks like I may not be able to do my task until this issue is resolved. Please add support for flutter web in this library.

**Comment by nemoryoliver** [comment]
Patiently waiting

**Comment by johnareid54** [comment]
I am using Flutter (Channel beta, 1.21.0-9.1.pre, on Mac OS X 10.15.6 19G73, locale en) with path-provider 1.6.14, the following works on Android and IOS but fails when run on web.

`import 'package:flutter/material.dart';
import 'package:hive/hive.dart';
import 'package:path_provider/path_provider.dart' as path_provider;

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  final appDocumentDir = await path_provider.getApplicationDocumentsDirectory();
  Hive.init(appDocumentDir.path);
  print(appDocumentDir.path);
  runApp(MyApp());
}`

**Comment by ditman** [comment]
@johnareid54 path_provider is not supported at all in web. Also, there's no standard concept "filesystem" or "path" on web browsers, so we'll first need to figure that out before we can implement a path_provider plugin.

**Comment by johnareid54** [comment]
@ditman  **path_provider is not supported at all in web.**
Yes after a few hours reading last night, I am beginning to understand the problem, but would it not be better to return null rather than not returning anything? I was looking to use a local database such as hive or sembast (which produces a different pub package for web), but both have workarounds for the web, but it means having to have one version for the web and another for mobile, which is far from ideal and goes against the overall concept of flutter.

**Comment by ditman** [comment]
> would it not be better to return null rather than not returning anything?

@johnareid54 The [default behavior](https://github.com/flutter/plugins/blob/master/packages/path_provider/path_provider_platform_interface/lib/path_provider_platform_interface.dart) for unimplemented plugins is to throw an `UnimplementedError` (which should be catchable from the App).

How is `return null` more helpful than something that you can `try/catch`? (Honest question, the plugin is already federated, so implementing a "return null" version would be quite fast to do)

**Comment by Katekko** [comment]
> @johnareid54 path_provider is not supported at all in web. Also, there's no standard concept "filesystem" or "path" on web browsers, so we'll first need to figure that out before we can implement a path_provider plugin.

I think the better way to do something is open a default modal like when you are downloading something, will popop up a modal for you save the download in any place. We can use the same concept. Instead pick the path it self just will open some modal for us, for picking ours self

**Comment by ditman** [comment]
> I think the better way to do something is open a default modal like when you are downloading something, will popop up a modal for you save the download in any place.

@Katekko the `path_provider` plugin returns paths (as strings) to common locations on the filesystem. There's no guarantee that the user of the plugin is using those paths to write files (you may be loading files, or just listing files and directories, or checking if a location exists, or doing anything else with the filesystem at that location...)

This experimental API might do what you describe [File System Access API](https://web.dev/file-system-access), but it's only supported in Chrome. There's a plugin in development that might match that API more closely: [file_selector](https://github.com/flutter/plugins/tree/master/packages/file_selector/file_selector_platform_interface), but again, it doesn't have a similar API to `path_provider`.

**Comment by mukkulsingh** [comment]
Patiently waiting for the web support

**Comment by stuartmorgan** [comment]
Please see the discussion above; the issue in not that nobody has implemented it, it's that there does not appear to be any meaningful way to implement this plugin on web.

Suggestions are welcome if anyone has concrete examples of how they believe it should be implemented.

**Comment by poka-IT** [comment]
> Suggestions are welcome if anyone has concrete examples of how they believe it should be implemented.

Is it not possible to use a virtual path in local storage instead of real filesystem path ?

Ok, the data will not be the same for sure, can't access to filesystem, but it could resolve things for people juste using `getApplicationDocumentsDirectory()` to store specifics datas of the app ?

**Comment by stuartmorgan** [comment]
I'm not sure what you mean by "a virtual path in local storage". Can you show an example of how you would construct a `Directory` object to return on the web, and how that would interact with existing code?

Or are you proposing a breaking change to `path_provider` that fundamentally alters how it works?

**Comment by hasanm08** [comment]
can you please fix this issue? I need getTemporaryDirectory in the web platform to cache my stream service data

**Comment by stuartmorgan** [comment]
Please read the discussion here before commenting, especially [this comment](https://github.com/flutter/flutter/issues/45296#issuecomment-804211869).

**Comment by hasanm08** [comment]
> e discussion above; the issue is not that nobody has implemented it, it's that **there doe**

new browsers have temp memory for each tab check youtube.com for a cache-based web streaming service

**Comment by stuartmorgan** [comment]
There is a difference between an API that provides storage and an API that returns a path that allows creating files, as has already been discussed above.

Again, if someone believes this can be implemented for web, please provide a **specific API** that would allow implementing it.

**Comment by or-adar** [comment]
> There is a difference between an API that provides storage and an API that returns a path that allows creating files, as has already been discussed above.
> 
> Again, if someone believes this can be implemented for web, please provide a **specific API** that would allow implementing it.

To be clear, the current security model of browsers only allows read & writes to go through a sandboxed file system or by prompting a file picker which grants access to the selected files/directories.
This behaviour is supported when using the [File System Access API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API), however this API isn't standardized, 
meaning that not all browsers support it (and not on all platforms).
So to acheive a broad support we are left with creating an input tag, like so:
```html
<input type="file" />
```

**Comment by stuartmorgan** [comment]
> So to acheive a broad support we are left with creating an input tag, like so:
> 
> ```
> <input type="file" />
> ```

That is not compatible with path_provider; asking the user to pick a path is not at the same thing as programatically accessing a known path. It's a completely different use case, which is in fact covered by a completely different plugin (`file_selector`).

This is not a generic request for access to some files on web, it's a request for an implementation of `path_provider`. That means that any proposed solution must be **compatible with the `path_provider` API and usage model**.

**Comment by or-adar** [comment]
> > So to acheive a broad support we are left with creating an input tag, like so:
> > ```
> > <input type="file" />
> > ```
> 
> That is not compatible with path_provider; asking the user to pick a path is not at the same thing as programatically accessing a known path. It's a completely different use case, which is in fact covered by a completely different plugin (`file_selector`).
> 
> This is not a generic request for access to some files on web, it's a request for an implementation of `path_provider`. That means that any proposed solution must be **compatible with the `path_provider` API and usage model**.

I get what you're trying to say.
In that case, [File System Access API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API) might help with the implementation of some methods, like getting locations of well-known directories: downloads, documents etc.
However, as I mentioned before, this API is not standardized, so it won't work on any browser.
And also, since prompting some sort of a file picker to receive access from a user is inevitable it will require to maintain another plugin/package that will be used to "mount" any well-known directory (using File System Access API) in order to enable methods in `path_provide` to refernce them later on.
Nevertheless, I don't know if it is something worth implementing given all the cons and the fact that we have to trust our user in picking the right directory to access.
I do agree that it sounds a lot of effort to support something that is quite similar to things you can acheive using `file_selector`.

**Comment by ditman** [comment]
@or-adar I've been looking at the API you mentioned since around it was made available (see this https://github.com/flutter/flutter/issues/45296#issuecomment-706292593, for example), and as you found out, it relies on the user picking the correct directory for their app every time.

The user picking the correct directory every time is probably not the biggest problem, though!

The current usage of the results of calling `path_provider` methods (that return Strings) is through `dart:io` Filesystem APIs that don't work **at all** on the web, so even if we implemented `path_provider` backed by the FileSystem Access API, your app code would still need different code depending on where you want to read/write your files: on mobile (using dart:io) or on web (using dart:html). 

**Since you have to write different code anyways once you receive a "selected path", you might as well use a different plugin altogether (`file_selector` + `cross_file` on the web and `path_provider` + `dart:io` on mobile).**

_(Maybe `path_provider` could be somehow made compatible with `package:cross_file` or similar, but at that point, its API would be so different that every user of `path_provider` would need to rewrite the parts of their app that used dart:io before...)_

Anyway, here's the problems we're facing to support this plugin on the web platform:

**From a web standards standpoint:**

* Paths on the web didn't mean anything until the File System Access API showed up in October 2020.
  * Very low browser support ([caniuse.com](https://caniuse.com/native-filesystem-api)).
  * Impossible to write a client-side replacement / shim for this.
* It's unlikely that Firefox will implement the API as is currently proposed, since it is considered **harmful** (https://github.com/mozilla/standards-positions/pull/545)

**From a Flutter perspective:**

* Even if `path_provider` worked, the code to actually use files from the selected paths would be completely different between mobile/web (because of `dart:io` vs `dart:html`)
  * If we provided a cross-platform way of doing File access, everybody would need to rewrite their apps to adopt it.

----

### TL;DR: Not so simple!

**Comment by hasanm08** [comment]
> @or-adar I've been looking at the API you mentioned since around it was made available (see this [#45296 (comment)](https://github.com/flutter/flutter/issues/45296#issuecomment-706292593), for example), and as you found out, it relies on the user picking the correct directory for their app every time.
> 
> The user picking the correct directory every time is probably not the biggest problem, though!
> 
> The current usage of the results of calling `path_provider` methods (that return Strings) is through `dart:io` Filesystem APIs that don't work **at all** on the web, so even if we implemented `path_provider` backed by the FileSystem Access API, your app code would still need different code depending on where you want to read/write your files: on mobile (using dart:io) or on web (using dart:html).
> 
> **Since you have to write different code anyways once you receive a "selected path", you might as well use a different plugin altogether (`file_selector` + `cross_file` on the web and `path_provider` + `dart:io` on mobile).**
> 
> _(Maybe `path_provider` could be somehow made compatible with `package:cross_file` or similar, but at that point, its API would be so different that every user of `path_provider` would need to rewrite the parts of their app that used dart:io before...)_
> 
> Anyway, here's the problems we're facing to support this plugin on the web platform:
> 
> **From a web standards standpoint:**
> 
> * Paths on the web didn't mean anything until the File System Access API showed up in October 2020.
>   
>   * Very low browser support ([caniuse.com](https://caniuse.com/native-filesystem-api)).
>   * Impossible to write a client-side replacement / shim for this.
> * It's unlikely that Firefox will implement the API as is currently proposed, since it is considered **harmful** ([Update File System Access (harmful) mozilla/standards-positions#545](https://github.com/mozilla/standards-positions/pull/545))
> 
> **From a Flutter perspective:**
> 
> * Even if `path_provider` worked, the code to actually use files from the selected paths would be completely different between mobile/web (because of `dart:io` vs `dart:html`)
>   
>   * If we provided a cross-platform way of doing File access, everybody would need to rewrite their apps to adopt it.
> 
> ### TL;DR: Not so simple!

If you merge dart io and dart html many apps and packages will support web too
Maybe some changes needed but it's worth it

**Comment by isanecat** [comment]
I understand that there is non web support for path_provider. Most of the web apps use cloud storage functionality. Since we have no path on local , creating the pdf on cloud storage may be a solution?

**Comment by yingming25** [comment]
I find web  File System Access API belowd:
https://web.dev/file-system-access/
https://css-tricks.com/getting-started-with-the-file-system-access-api/

Not sure can help.

**Comment by ditman** [comment]
@yingming25 I've been following that API for a while, but as of now, it's not very well supported:

[![Can I use File System Access API?](https://caniuse.bitsofco.de/image/native-filesystem-api.png)](https://caniuse.com/native-filesystem-api)

**Comment by AymericLeFeyer** [comment]
Currently use PhraseApp in my app. It fails for the web version due to path_provider

**Comment by mateisilviu** [comment]
I have a solution for Firebase storage, my issue was with PDF file, with this snippet you could download or preview PDF in browser.
[Flutter Web download pdf as files from Firebase storage](https://gist.github.com/mateisilviu/5ef432baf964942a9e98d01fd51a8257 )

**Comment by ditman** [comment]
It seems that both Chrome and Webkit are implementing the "Origin Private File System":

* [Origin Private File System](https://webkit.org/blog/12257/the-file-system-access-api-with-origin-private-file-system/)
* [4. Accessing the Origin Private File System](https://wicg.github.io/file-system-access/#sandboxed-filesystem)

The API above seems to provide files + directories that are persisted in some form of "private storage" that the browser keeps hidden, like the sandboxed storages of mobile apps.

It is unclear to me what's the quota, or what happens if a user "clears their local storage", if you're going to be able to "pick" a file from the private storage, or how to make this work with `XFile` (maybe we need a major version bump to that package), but it looks promising!

(Firefox seems to be [working on it too](https://bugzilla.mozilla.org/show_bug.cgi?id=1748667).)

**Comment by junhaow1** [comment]
> Waiting for the web support. I am implementing "export as csv" functionality on my app on web, Android and iOS. Web fails because there is no support.

found same problem. did u solve it mate? :)

**Comment by loic-hamdi** [comment]
Any update?

**Comment by ramsunvtech** [comment]
Its seems like issue is still exist. path_provider should support [indexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) / LocalStorage

**Comment by loic-hamdi** [comment]
> @yingming25 I've been following that API for a while, but as of now, it's not very well supported:
> 
> [![Can I use File System Access API?](https://camo.githubusercontent.com/04f9b866f90509d0a298eb24dc08353de9812f863faa5d9023b6cf80fbbd61ba/68747470733a2f2f63616e697573652e626974736f66636f2e64652f696d6167652f6e61746976652d66696c6573797374656d2d6170692e706e67)](https://caniuse.com/native-filesystem-api)
> 
> This may be another way to go: https://developer.mozilla.org/en-US/docs/Web/API/FileSystem
> 
> (But again, the above needs to be supported by dart or a plugin that collaborates with the path_provider, otherwise it won't make sense).




It seems that from the link you shared, the FileSystem API implementation made excellent progress.

<img width="784" alt="Screenshot 2023-03-16 at 9 55 59 AM" src="https://user-images.githubusercontent.com/3080026/225657663-396ab77b-1c61-4c06-9809-3092e71d55d4.png">

**Comment by ditman** [comment]
The most promising API at the moment is OPFS:

https://caniuse.com/native-filesystem-api

**Comment by loic-hamdi** [comment]
This feature is so essential if we want Flutter to be a credible solution for web, it is hard to believe that the ticket is open since 2019.

Is there any alternative solution or workaround?

**Comment by ubershmekel** [comment]
I was reading a file in my flutter web project and got this error:

```
Error: Unsupported operation: _Namespace
```

I searched the web and found an answer that the browser can't read files. The solution has an `if` statement that [checks if we're on the web or not](https://stackoverflow.com/a/67848001/177498). But I assumed there was a multi-platform fluttery way of doing it.

Then I found https://docs.flutter.dev/cookbook/persistence/reading-writing-files which pointed me to this issue:

> Note: This recipe doesn’t work with web apps at this time. To follow the discussion on this issue, check out flutter/flutter [issue #45296](https://github.com/flutter/flutter/issues/45296).

But all I really wanted was this line:

`return await rootBundle.loadString('assets/my_text.txt');`

Which I found in this [answer](https://stackoverflow.com/a/54133627/177498).

I could open another issue complaining that the original error message is cryptic. But perhaps adding the information in this issue is enough?

**Comment by gokhanmeteerturk** [comment]
Basic OPFS features are now supported by all major browsers. Isn't it enough to add support for at least the getApplicationDocumentsDirectory?

**Comment by ditman** [comment]
There's a good explainer here:

* https://web.dev/articles/origin-private-file-system

We may be able to improve some features of cross_file (like Save As...) using the API.

Note that the OPFS is private and invisible to the user, so anything that needs to be downloaded to the public user FS needs to be "saved as".

**Comment by MilkBiscuit** [comment]
Wow, nearly five years and it's still open...

**Comment by gokhanmeteerturk** [comment]
For anyone watching this with the hopes of having an easy alternative to something like an "Application Support" directory on web, [file_system_access_api](https://pub.dev/packages/file_system_access_api) package by @poirierlouis gives you access to OPFS but is experimental.

**Comment by poirierlouis** [comment]
I'll add that the library in itself works fine. What's marked as experimental is the specification. But OPFS is widely implemented in major browsers iirc.

**Comment by ditman** [comment]
OPFS is great, and it could be used to simplify a lot of code in other web plugins that show open/save pickers (basically everything backed by `package:cross_file`, like `image_picker` and `file_selector`).

The problem with `path_provider` is that it is currently backed by `dart:io` `Directory` objects that are not available on the web.

So before `path_provider_web` can happen, we need something that abstracts directory operations on the web (imagine a `XDirectory` class in `cross_file`).

(PS: `file_system_access_api` wouldn't be usable for `flutter/packages` as is right now, because it's implemented with `dart:html` and it won't compile to WASM, however `package:web` offers all of the MDN documented classes)

Anyway, the technical hurdles seem doable, and OPFS is exciting and can improve a lot of the flutter packages that handle `XFile` objects.

---

I have a question though, what to do with the directories that break the expectation of what the app developer expects on the users' machines? From the [list of directories](https://pub.dev/packages/path_provider#supported-platforms-and-paths), only the ones that are internal to the application would work "as expected" (by your app). Downloads would be impossible to keep in sync with the real Downloads directory on the users' desktop, or we should mark it with a fat ❌.

Would that be acceptable?

**Comment by poirierlouis** [comment]
> The problem with path_provider is that it is currently backed by dart:io Directory objects that are not available on the web.

Regarding this issue, I started a [branch](https://github.com/poirierlouis/file_system_access_api/compare/master...io) a while back, to implement Directory/File from `dart:io`, using `file_system_access_api`. But it is limited (maybe not even tested), as only asynchronous is supported. Therefore some methods throws unimplemented errors. In some way cross-platform + web could be used this way, but only with async mode. Only saying in case it can brings ideas...

**Comment by ditman** [comment]
> > The problem with path_provider is that it is currently backed by dart:io Directory objects that are not available on the web.
> 
> Regarding this issue, I started a [branch](https://github.com/poirierlouis/file_system_access_api/compare/master...io) a while back, to implement Directory/File from `dart:io`, using `file_system_access_api`. But it is limited (maybe not even tested), as only asynchronous is supported. Therefore some methods throws unimplemented errors. In some way cross-platform + web could be used this way, but only with async mode. Only saying in case it can brings ideas...

Ah yes, we have [`package:cross_file`](https://pub.dev/packages/cross_file) which is similar to what you were doing there, and we use it across all packages that handle files in `flutter/packages`. We just never needed "`XDirectory`" so far.

**Comment by saibotma** [comment]
> Would that be acceptable?

Yes
